frontend/rust-lib/flowy-search/src/folder/indexer.rs

```diff
@@ -0,0 +1,360 @@
+use collab_folder::{View, ViewIcon, ViewIndexContent, ViewLayout};
+use flowy_error::{FlowyError, FlowyResult};
+use flowy_user::services::authenticate_user::AuthenticateUser;
+use lib_dispatch::prelude::af_spawn;
+use strsim::levenshtein;
+use tantivy::{
+  collector::TopDocs, directory::MmapDirectory, doc, query::QueryParser, Index, IndexReader,
+  IndexWriter, Term,
+};
+
+use crate::{
+  entities::SearchResultPB,
+  services::indexer::{IndexManager, IndexableData},
+};
+
+use super::{
+  entities::FolderIndexData,
+  schema::{FOLDER_ICON_TY_FIELD_NAME, FOLDER_ID_FIELD_NAME},
+};
+
+#[derive(Clone)]
+pub struct FolderIndexManager {
+  folder_schema: Option<FolderSchema>,
+  index: Option<Index>,
+  index_reader: Option<IndexReader>,
+}
+
+const FOLDER_INDEX_DIR: &str = "folder_index";
+
+impl FolderIndexManager {
+  pub fn new(auth_user: Weak<AuthenticateUser>) -> Self {
+    let authenticate_user = auth_user.upgrade();
+    let storage_path = match authenticate_user {
+      Some(auth_user) => auth_user.get_index_path(),
+      None => {
+        tracing::error!("FolderIndexManager: AuthenticateUser is not available");
+        return FolderIndexManager::empty();
+      },
+    };
+
+    let index_path = storage_path.join(Path::new(FOLDER_INDEX_DIR));
+    if !index_path.exists() {
+      let res = fs::create_dir_all(&index_path);
+      if let Err(e) = res {
+        tracing::error!(
+          "FolderIndexManager failed to create index directory: {:?}",
+          e
+        );
+        return FolderIndexManager::empty();
+      }
+    }
+
+    let dir = MmapDirectory::open(index_path);
+    if let Err(e) = dir {
+      tracing::error!("FolderIndexManager failed to open index directory: {:?}", e);
+      return FolderIndexManager::empty();
+    }
+
+    let folder_schema = FolderSchema::new();
+    let index_res = Index::open_or_create(dir.unwrap(), folder_schema.clone().schema);
+    if let Err(e) = index_res {
+      tracing::error!("FolderIndexManager failed to open index: {:?}", e);
+      return FolderIndexManager::empty();
+    }
+
+    let index = index_res.unwrap();
+    let index_reader = index.reader();
+    if let Err(e) = index_reader {
+      tracing::error!(
+        "FolderIndexManager failed to instantiate index reader: {:?}",
+        e
+      );
+      return FolderIndexManager::empty();
+    }
+
+    Self {
+      folder_schema: Some(folder_schema),
+      index: Some(index),
+      index_reader: Some(index_reader.unwrap()),
+    }
+  }
+
+  pub fn index_all_views(&self, views: Vec<View>) {
+    let indexable_data = views
+      .iter()
+      .map(|view| IndexableData {
+        id: view.id.clone(),
+        data: view.name.clone(),
+        icon: view.icon.clone(),
+        layout: view.layout.clone(),
+      })
+      .collect();
+
+    let _ = self.index_all(indexable_data);
+  }
+
+  fn index_all(&self, indexes: Vec<IndexableData>) -> Result<(), FlowyError> {
```

Comment:
make sure index_all is called on tokio::task::spawnblocking if it would use too much cpu time